✨ ADR for Runtime/engine/host/environment support and CI (#365)
expressjs/discussions: docs/adr/environments-and-ci.md

# ADR 35580f0d-f429-412b-acef-83655e3cab11: Runtime/engine/host/environment support and CI

## Status

Proposed

## Submitters

- @ctcpip

## Decision Owners

- @expressjs/express-tc

## Context

> Express and its libraries were specifically designed to run with Node.js (V8). While some of our libraries can run in other environments (e.g. runtimes, engines, browsers), they are not officially supported. Consequently, our CI systems do not include other environments as part of their testing workflows.

Author: resolving per agreement with Blake

Several points were raised during the discussion:

- Cost of running additional CI vs the likelihood of detecting a problem
- Introducing maintenance overhead and possible coupling to other environments' development lifecycle
- No JS engine implements ECMAScript 100% correctly; thus, claiming "ES2015 support" does not guarantee correctness across all environments.
- Engine regressions or language edge cases could break functionality in unpredictable ways that are not practical for us to monitor across all environments.

## Decision

- We will **not** add non-Node.js environment testing to our CI pipelines.
- CI will continue to run only against supported Node.js versions.

> - Unofficial support for other environments may exist, but we do not guarantee correctness or compatibility across those environments.

Author: I think the latest version of this line captures this sufficiently, but please say if not.  Marking as resolved per our discussion in slack

# ADR 35580f0d-f429-412b-acef-83655e3cab11: Runtime/engine/host/environment support and CI

## Status

Proposed

## Submitters

- @ctcpip

## Decision Owners

- @expressjs/express-tc

## Context

> Express and its libraries were specifically designed to run with Node.js (V8). While some of our libraries can run in other environments (e.g. runtimes, engines, browsers), they are not necessarily supported in all environments. Consequently, our CI systems do not include other environments as part of their testing workflows.

Author: yeah, happy to capture that if you have a concrete suggestion, but this is just some background context to set the stage, and don't want to risk comprehension with potentially excessive qualification and detail at this point

> Express and its libraries were specifically designed to run with Node.js (V8). While some of our libraries can run in other environments (e.g. runtimes, engines, browsers), they are not necessarily supported in all environments. Consequently, our CI systems do not include other environments as part of their testing workflows.

Reviewer: > Express and its libraries were originally designed to run on Node.js (V8). While some libraries can run in other environments (e.g. runtimes, engines, browsers), intentionally or not, they are not necessarily supported everywhere. Consequently, our CI systems do not include other environments as part of their testing workflows.

Honestly it looks fine already, and I think the comment about maintenance below covers the realities of `path-to-regexp`.

Several points were raised during the discussion:

- Cost of running additional CI vs the likelihood of detecting a problem
- Introducing maintenance overhead and possible coupling to other environments' development lifecycle
- No JS engine implements ECMAScript 100% correctly; thus, claiming "ES2015 support" does not guarantee correctness across all environments.
- Environment regressions or language edge cases could break functionality in unpredictable ways that are not practical for us to monitor across all environments.

## Decision

> - We will **not** add non-Node.js environment testing to our CI pipelines.

Author: adding an exception sub-bullet to reflect

> - We will **not** add non-Node.js environment testing to our CI pipelines.

Author: pushed, marking resolved, but please reopen if not sufficient

- CI will continue to run only against supported Node.js versions.

> - Support for other environments may exist, but we do not guarantee correctness or compatibility across all environments.

Reviewer: I think "all environments" is already clear, maybe I just need confirmation on how I read it. I can imagine someone coming here and reading it as "Safari might be supported but compatibility is not guaranteed" vs "Safari might be supported and guaranteed, but <obscure JS engine> is not".

> - Support for other environments may exist, but we do not guarantee correctness or compatibility across all environments.

Author: hmmm.. perhaps a more general approach:

> Support for other environments may exist, per maintainer discretion, but we do not guarantee support across all environments.

> - Support for other environments may exist, but we do not guarantee correctness or compatibility across all environments.

Reviewer: Sounds perfect 👍

  - Some libraries, particularly language-only libraries which do not require non-language APIs, strive to support as many environments as possible.
    - Nonetheless, support is not guaranteed across every possible environment, and is provided on a best-effort basis.

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Author: applying suggestion with a small tweak

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Author: I'm making the change but I am apprehensive about "include information about supported environments".  I explicitly do not want to be in the business of listing specific runtimes, engines, etc.

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Reviewer: I agree. I wanted to find a way to capture "this thing should run on Safari 10 just fine" without promising the world, so if there's a better way to phrase it that'd be great. Not everyone reads ES2015 or "uses generators and classes" and goes "ah, Safari 10".

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Reviewer: Working backward, I think a sentence I'd write for `path-to-regexp` is something like:

> Supports ES2015 runtimes such as Node X+, Deno, Bun, Chrome, Safari X+, etc.

That's already a mouthful but hopefully it makes sense. I do want someone to legitimately feel happy knowing Safari issues would be fixed if they found an issue and not closed because ES blah blah.

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Author: yeah, it's fine.  the risk here is that we also don't want to be overly prescriptive on content. but it would be great if someone had an idea to capture this spirit better.  maybe something will come to mind later

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Author: > Supports ES2015 runtimes such as Node X+, Deno, Bun, Chrome, Safari X+, etc.

hmmm, this is precisely what I'd like to avoid  😅

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Reviewer: Perhaps it's more "general compatibility information, e.g. ECMAScript version, or environment requirements"? You're right, I shouldn't enumerate the environments but instead want a general statement about what environments should work.

I don't have a solid statement that I want to actually work backward from here. I'm trying to think of these two cases:

1. A simple no-environment specific library that may want to say "supports ES2015 and strives to support the last X years of popular browsers".
2. A library that supports X and Y explicitly, e.g. by using native packages such as `node:http`.

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Reviewer: I was overthinking the "but people might not know what environments support ES2015".

>   - Libraries will not explicitly list all supported environments; they may, however, state general compatibility information, e.g. ECMAScript version.

Author: you could always link to the compat table. e.g, for ES2015/ES6:  https://compat-table.github.io/compat-table/es6/

> - If issues are reported for other environments, maintainers may investigate at their discretion, but no automated validation or regression testing infrastructure will be built for them.

Author: ok, sure.  I guess this is already covered anyway by the other clause about CI

> - If issues are reported for other environments, maintainers may investigate at their discretion, but no automated validation or regression testing infrastructure will be built for them.

Author: also changed 'may' to 'will' since 'discretion' is already a qualifier

> - If issues are reported for other environments, maintainers may investigate at their discretion, but no automated validation or regression testing infrastructure will be built for them.

Author: pushed

# ADR 35580f0d-f429-412b-acef-83655e3cab11: Runtime/engine/host/environment support and CI

## Status

Proposed

## Submitters

- @ctcpip

## Decision Owners

- @expressjs/express-tc

## Context

> Express and its libraries were specifically designed to run with Node.js (V8). While some of our libraries can run in other environments (e.g. runtimes, engines, browsers), they are not necessarily supported in all environments. Consequently, our CI systems do not include other environments as part of their testing workflows.

Reviewer 2: ```suggestion
Express and its libraries were specifically designed to run with Node.js (V8). While some of our libraries can run in other environments (e.g. runtimes, engines, browsers), they are not necessarily supported in all environments—that is, we do not test against them, optimize for them, or guarantee compatibility. Consequently, our CI systems do not include other environments as part of their testing workflows.
```

> Express and its libraries were specifically designed to run with Node.js (V8). While some of our libraries can run in other environments (e.g. runtimes, engines, browsers), they are not necessarily supported in all environments. Consequently, our CI systems do not include other environments as part of their testing workflows.

Reviewer 3: To disambiguate a bit from "version" and hopefully make it more clear why that is relevant in this ADR:

```suggestion
Express and its libraries were specifically designed to run with Node.js (and the V8 JavaScript Engine). While some of our libraries can run in other environments (e.g. runtimes, engines, browsers), they are not necessarily supported in all environments. Consequently, our CI systems do not include other environments as part of their testing workflows.
```

Several points were raised during the discussion:

- Cost of running additional CI vs the likelihood of detecting a problem
- Introducing maintenance overhead and possible coupling to other environments' development lifecycle
- No JS engine implements ECMAScript 100% correctly; thus, claiming "ES2015 support" does not guarantee correctness across all environments.
- Environment regressions or language edge cases could break functionality in unpredictable ways that are not practical for us to monitor across all environments.

## Decision

- We will **not** add non-Node.js environment testing to our CI pipelines.

>   - Exceptions may be made iff there is strongly compelling, project-aligned justification.

Reviewer 4: ```suggestion
  - Exceptions may be made if there is strongly compelling, project-aligned justification.
```

Author: https://en.wiktionary.org/wiki/iff   🙂

Reviewer 3: Ok I am confused, is this intentional or an etymology joke 😆 ?

Either way, lets merge the suggestion right?

- CI will continue to run only against supported Node.js versions.
- Support for other environments may exist, per maintainer discretion, but we do not guarantee support across all environments.

>   - Some libraries, particularly language-only libraries which do not require non-language APIs, strive to support as many environments as possible.

Reviewer 3: Hoping this language may be more clear? I did not understand at first what "language-only" meant and it took a few reads for me to be sure. 

```suggestion
  - Some libraries, particularly libraries which do not require non-ECMA APIs (Node.js apis, WHATWG apis, etc), strive to support as many environments as possible.
```

> - Support for other environments may exist, per maintainer discretion, but we do not guarantee support across all environments.
>   - Some libraries, particularly language-only libraries which do not require non-language APIs, strive to support as many environments as possible.
>     - Nonetheless, support is not guaranteed across every possible environment, and is provided on a best-effort basis.

Reviewer 2: ```suggestion
- Some Express libraries may work in other environments (e.g., Bun, Deno, Cloudflare Workers), but we do not guarantee compatibility or prioritize testing and development for them. Support outside Node.js is best-effort and may vary between packages, depending on maintainer interest and alignment with project goals.
  - Some libraries, particularly language-only libraries which do not require non-language APIs, strive to support as many environments as possible.
```

Author: intent is to deliberately avoid mentioning specific environments

Author: https://github.com/expressjs/discussions/pull/365#discussion_r2067487965

Reviewer 2: edited out the mention

  - Libraries may state general compatibility information, e.g. ECMAScript version, and possibly information about supported environments, but will not explicitly list all supported environments.

> - If issues are reported for other environments, maintainers will investigate at their discretion.

Reviewer 3: ```suggestion
- If issues are reported for other environments, maintainers may investigate at their discretion.
```

## Rationale

- **Alternatives Considered:**
  - **Add support for additional environments in CI**: Rejected due to complexity and minimal return on value.

- **Pros and Cons**:

  **Pros**:
  - Keeps CI lightweight and maintainable
  - Avoids implicit endorsement of non-Node.js environments
  - Maintains focus on Express's design goals and core user base

  **Cons**:
  - Some users may misinterpret lack of other environment testing as non-support
  - Manual verification may be required when bugs are reported in other environments

- **Why is this decision the best option?**
  - It balances clarity, project focus, and contributor/maintainer effort. It avoids premature optimization or expanding scope into formal environment support.

## Consequences

- **Positive Impact**:
  - Reduced CI runtime and maintenance burden
  - Clearer expectations about what we support and test

- **Negative Impact**:

>   - Users of alternate environments may find compatibility issues undetected until runtime

Reviewer 2: ```suggestion
  - Users of alternate environments (e.g., Bun, Deno, Cloudflare Workers) may find that Express libraries appear to work but fail under certain conditions due to untested APIs or platform differences. This may result in runtime issues that are not prioritized for triage or resolution unless clearly aligned with the project’s goals.
```
